libsrc/arithmetic/man3/im_lintra.3
.TH IM_LINTRA 3 "24 April 1991"
.SH NAME
im_lintra, im_lintra_vec \- performs a linear transformation on an image
.SH SYNOPSIS
.B #include <vips/vips.h>

.B int im_lintra_vec(n, a, in, b, out)
.br
.B int n;
.br
.B double *a, *b;
.br
.B IMAGE *in, *out;

.B int im_lintra(a, in, b, out)
.br
.B double a, b;
.br
.B IMAGE *in, *out;

.SH DESCRIPTION
.B im_lintra_vec(3)
performs a linear transform on image in, that is, it calculates

  out = a * in + b

.B a
and 
.B b
are vectors, ie. arrays of constants of length
.B n.
If 
.B in
has one band, then the vectors may be any length and the output image will
have the same number of bands as the length of the vector. If 
.B in
has many bands, then the vector must be length 1, or have the same length as
the number of bands in the image. 

If the input format is one of the integer types
then output is float.  In all other cases the
output is the same as the input.

.B im_lintra(3)
is a convenience function which calls
.B im_lintra_vec(3)
with a vector of length 1.

.SH RETURN VALUE
The function returns 0 on success and -1 on error.
.SH BUGS
The function does not check for under/overflow
.SH SEE\ ALSO
im_exptra(3), im_logtra(3)
.SH AUTHOR
N. Dessipris \- 24/04/1991
.br
J. Cupitt (rewrite) \- 21/7/93


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/arithmetic/im_lintra.c
/* @(#) Pass an image through a linear transform - ie. out = in*a + b. Output
 * @(#) is always float for integer input, double for double input, complex for
 * @(#) complex input and double complex for double complex input.
 * @(#)
 * @(#) int 
 * @(#) im_lintra( a, in, b, out )
 * @(#) IMAGE *in, *out;
 * @(#) double a, b;
 * @(#)
 * @(#) Returns 0 on success and -1 on error
 * @(#)
 *
 * Copyright: 1990, N. Dessipris, based on im_powtra()
 * Author: Nicos Dessipris
 * Written on: 02/05/1990
 * Modified on: 
 * 23/4/93 JC
 *	- adapted to work with partial images
 * 1/7/93 JC
 *	- adapted for partial v2
 * 7/10/94 JC
 *	- new IM_NEW()
 *	- more typedefs 
 * 9/2/95 JC
 *	- adapted for im_wrap...
 *	- operations on complex images now just transform the real channel
 * 29/9/95 JC
 *	- complex was broken
 * 15/4/97 JC
 *	- return(0) missing from generate, arrgh!
 * 1/7/98 JC
 *	- im_lintra_vec added
 * 3/8/02 JC
 *	- fall back to im_copy() for a == 1, b == 0
 * 10/10/02 JC
 *	- auug, failing to multiply imag for complex! (thanks matt)
 * 10/12/02 JC
 *	- removed im_copy() fallback ... meant that output format could change
 *	  with value :-( very confusing
 * 30/6/04
 *	- added 1 band image * n band vector case
 * 8/12/06
 * 	- add liboil support
 */

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <assert.h>

#include <vips/vips.h>
#include <vips/internal.h>

#ifdef HAVE_LIBOIL
#include <liboil/liboil.h>
#endif /*HAVE_LIBOIL*/

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Struct we need for im_generate().
 */
typedef struct {
	int n;			/* Number of bands of constants */
	double *a, *b;
} LintraInfo;

/* Define what we do for each band element type. Non-complex input, any
 * output.
 */
#define LOOP(IN, OUT) { \
	IN *p = (IN *) in; \
	OUT *q = (OUT *) out; \
	\
	for( x = 0; x < sz; x++ ) \
		q[x] = a * (OUT) p[x] + b; \
}

/* Complex input, complex output. 
 */
#define LOOPCMPLX(IN, OUT) {\
	IN *p = (IN *) in; \
	OUT *q = (OUT *) out; \
	\
	for( x = 0; x < sz; x++ ) { \
		q[0] = a * p[0] + b; \
		q[1] = a * p[1]; \
		q += 2; \
		p += 2; \
	} \
}

#ifdef HAVE_LIBOIL
/* Process granularity.
 */
#define CHUNKS (1000)

/* d[] = s[] * b + c, with liboil
 */
static void
lintra_f32( float *d, float *s, int n, float b, float c )
{
	float buf[CHUNKS];
	int i;

	for( i = 0; i < n; i += CHUNKS ) {
		oil_scalarmultiply_f32_ns( buf, s, 
			&b, IM_MIN( CHUNKS, n - i ) );
		oil_scalaradd_f32_ns( d, buf, 
			&c, IM_MIN( CHUNKS, n - i ) );

		s += CHUNKS;
		d += CHUNKS;
	}
}
#endif /*HAVE_LIBOIL*/

/* Lintra a buffer, 1 set of scale/offset.
 */
static int
lintra1_gen( PEL *in, PEL *out, int width, IMAGE *im, LintraInfo *inf )
{	
	double a = inf->a[0];
	double b = inf->b[0];
	int sz = width * im->Bands;
	int x;

	/* Lintra all input types.
         */
        switch( im->BandFmt ) {
        case IM_BANDFMT_UCHAR: 		LOOP( unsigned char, float ); break;
        case IM_BANDFMT_CHAR: 		LOOP( signed char, float ); break; 
        case IM_BANDFMT_USHORT: 	LOOP( unsigned short, float ); break; 
        case IM_BANDFMT_SHORT: 		LOOP( signed short, float ); break; 
        case IM_BANDFMT_UINT: 		LOOP( unsigned int, float ); break; 
        case IM_BANDFMT_INT: 		LOOP( signed int, float );  break; 
        case IM_BANDFMT_FLOAT: 		
#ifdef HAVE_LIBOIL
		lintra_f32( (float *) out, (float *) in, sz, a, b );
#else /*!HAVE_LIBOIL*/
		LOOP( float, float ); 
#endif /*HAVE_LIBOIL*/
		break; 

        case IM_BANDFMT_DOUBLE:		LOOP( double, double ); break; 
        case IM_BANDFMT_COMPLEX:	LOOPCMPLX( float, float ); break; 
        case IM_BANDFMT_DPCOMPLEX:	LOOPCMPLX( double, double ); break;

        default:
		assert( 0 );
        }

	return( 0 );
}

/* Define what we do for each band element type. Non-complex input, any
 * output.
 */
#define LOOPN(IN, OUT)\
	{\
		IN *p = (IN *) in;\
		OUT *q = (OUT *) out;\
		\
		for( i = 0, x = 0; x < width; x++ )\
			for( k = 0; k < nb; k++, i++ )\
				q[i] = a[k] * (OUT) p[i] + b[k];\
	}

/* Complex input, complex output. 
 */
#define LOOPCMPLXN(IN, OUT)\
	{\
		IN *p = (IN *) in;\
		OUT *q = (OUT *) out;\
		\
		for( x = 0; x < width; x++ ) \
			for( k = 0; k < nb; k++ ) {\
				q[0] = a[k] * p[0] + b[k];\
				q[1] = a[k] * p[1];\
				q += 2;\
				p += 2;\
			}\
	}

/* Lintra a buffer, n set of scale/offset.
 */
static int
lintran_gen( PEL *in, PEL *out, int width, IMAGE *im, LintraInfo *inf )
{
	double *a = inf->a;
	double *b = inf->b;
	int nb = im->Bands;
	int i, x, k;

	/* Lintra all input types.
         */
        switch( im->BandFmt ) {
        case IM_BANDFMT_UCHAR: 		LOOPN( unsigned char, float ); break;
        case IM_BANDFMT_CHAR: 		LOOPN( signed char, float ); break; 
        case IM_BANDFMT_USHORT: 	LOOPN( unsigned short, float ); break; 
        case IM_BANDFMT_SHORT: 		LOOPN( signed short, float ); break; 
        case IM_BANDFMT_UINT: 		LOOPN( unsigned int, float ); break; 
        case IM_BANDFMT_INT: 		LOOPN( signed int, float );  break; 
        case IM_BANDFMT_FLOAT: 		LOOPN( float, float ); break; 
        case IM_BANDFMT_DOUBLE:		LOOPN( double, double ); break; 
        case IM_BANDFMT_COMPLEX:	LOOPCMPLXN( float, float ); break; 
        case IM_BANDFMT_DPCOMPLEX:	LOOPCMPLXN( double, double ); break;

        default:
		assert( 0 );
        }

	return( 0 );
}

/* 1 band image, n band vector.
 */
#define LOOPNV(IN, OUT) { \
	IN *p = (IN *) in; \
	OUT *q = (OUT *) out; \
	\
	for( i = 0, x = 0; x < width; x++ ) { \
		OUT v = p[x]; \
		\
		for( k = 0; k < nb; k++, i++ ) \
			q[i] = a[k] * v + b[k]; \
	} \
}

#define LOOPCMPLXNV(IN, OUT) { \
	IN *p = (IN *) in; \
	OUT *q = (OUT *) out; \
	\
	for( x = 0; x < width; x++ ) { \
		OUT p0 = p[0]; \
		OUT p1 = p[1]; \
		\
		for( k = 0; k < nb; k++ ) { \
			q[0] = a[k] * p0 + b[k]; \
			q[1] = a[k] * p1; \
			q += 2; \
		} \
		\
		p += 2; \
	} \
}

static int
lintranv_gen( PEL *in, PEL *out, int width, IMAGE *im, LintraInfo *inf )
{
	double *a = inf->a;
	double *b = inf->b;
	int nb = inf->n;
	int i, x, k;

	/* Lintra all input types.
         */
        switch( im->BandFmt ) {
        case IM_BANDFMT_UCHAR: 		LOOPNV( unsigned char, float ); break;
        case IM_BANDFMT_CHAR: 		LOOPNV( signed char, float ); break; 
        case IM_BANDFMT_USHORT: 	LOOPNV( unsigned short, float ); break; 
        case IM_BANDFMT_SHORT: 		LOOPNV( signed short, float ); break; 
        case IM_BANDFMT_UINT: 		LOOPNV( unsigned int, float ); break; 
        case IM_BANDFMT_INT: 		LOOPNV( signed int, float );  break; 
        case IM_BANDFMT_FLOAT: 		LOOPNV( float, float ); break; 
        case IM_BANDFMT_DOUBLE:		LOOPNV( double, double ); break; 
        case IM_BANDFMT_COMPLEX:	LOOPCMPLXNV( float, float ); break; 
        case IM_BANDFMT_DPCOMPLEX:	LOOPCMPLXNV( double, double ); break;

        default:
		assert( 0 );
        }

	return( 0 );
}

/* Linear transform n bands.
 */
int 
im_lintra_vec( int n, double *a, IMAGE *in, double *b, IMAGE *out )
{	
	LintraInfo *inf;
	int i;

	/* Check args.
	 */
	if( in->Coding != IM_CODING_NONE ) {
		im_error( "im_lintra_vec", _( "not uncoded" ) );
		return( -1 );
	}

	/* If n and bands differ, one of them must be one (and we multiplex
	 * the other).
	 */
	if( n != in->Bands && (n != 1 && in->Bands != 1) ) {
		im_error( "im_lintra_vec", 
			_( "not 1 or %d elements in vector" ), in->Bands );
		return( -1 );
	}

	/* Prepare output header.
	 */
	if( im_cp_desc( out, in ) )
		return( -1 );
	if( im_isint( in ) ) {
		out->Bbits = IM_BBITS_FLOAT;
		out->BandFmt = IM_BANDFMT_FLOAT;
	}
	if( in->Bands == 1 )
		out->Bands = n;

	/* Make space for a little buffer.
	 */
	if( !(inf = IM_NEW( out, LintraInfo )) || 
		!(inf->a = IM_ARRAY( out, n, double )) ||
		!(inf->b = IM_ARRAY( out, n, double )) )
		return( -1 );
	inf->n = n;
	for( i = 0; i < n; i++ ) {
		inf->a[i] = a[i];
		inf->b[i] = b[i];
	}

	/* Generate!
	 */
	if( n == 1 ) {
		if( im_wrapone( in, out, 
			(im_wrapone_fn) lintra1_gen, in, inf ) )
			return( -1 );
	}
	else if( in->Bands == 1 ) {
		if( im_wrapone( in, out, 
			(im_wrapone_fn) lintranv_gen, in, inf ) )
			return( -1 );
	}
	else {
		if( im_wrapone( in, out, 
			(im_wrapone_fn) lintran_gen, in, inf ) )
			return( -1 );
	}

	return( 0 );
}

/* Linear transform.
 */
int 
im_lintra( double a, IMAGE *in, double b, IMAGE *out )
{	
	return( im_lintra_vec( 1, &a, in, &b, out ) );
}
